vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt

```diff
@@ -139,88 +189,204 @@ class HolderAgent(
+                    status = validator.checkRevocationStatus(it.vc)
                 )
 
                 is SubjectCredentialStore.StoreEntry.SdJwt -> Holder.StoredCredential.SdJwt(
-                    it.vcSerialized, it.sdJwt, Validator.RevocationStatus.VALID // TODO validation check
+                    storeEntry = it,
+                    status = validator.checkRevocationStatus(it.sdJwt)
                 )
             }
         }
     }
 
-    /**
-     * Creates a [VerifiablePresentation] serialized as a JWT for all the credentials we have stored,
-     * that match [credentialSchemes] (if specified).
-     * Optionally filters by [requestedClaims] (e.g. in ISO or SD-JWT case).
-     *
-     * May return null if no valid credentials (i.e. non-revoked, matching attribute name) are available.
-     */
     override suspend fun createPresentation(
         challenge: String,
         audienceId: String,
-        credentialSchemes: Collection<ConstantIndex.CredentialScheme>?,
-        requestedClaims: Collection<String>?,
-    ): Holder.CreatePresentationResult? {
-        val credentials = subjectCredentialStore.getCredentials(credentialSchemes).getOrNull()
-            ?: return null
-                .also { Napier.w("Got no credentials from subjectCredentialStore for $credentialSchemes") }
-        val validVcCredentials = credentials
-            .filterIsInstance<SubjectCredentialStore.StoreEntry.Vc>()
-            .filter { validator.checkRevocationStatus(it.vc) != Validator.RevocationStatus.REVOKED }
-            .map { it.vcSerialized }
-        if (validVcCredentials.isNotEmpty()) {
-            return createPresentation(validVcCredentials, challenge, audienceId)
-        }
-        val validIsoCredential = credentials
-            .filterIsInstance<SubjectCredentialStore.StoreEntry.Iso>()
-            // no revocation check
-            .firstOrNull()
-        if (validIsoCredential != null) {
-            return createIsoPresentation(challenge, validIsoCredential, requestedClaims)
+        presentationDefinition: PresentationDefinition,
+    ): KmmResult<Holder.HolderResponseParameters> {
+        // an attempt to implement input evaluation and submission according to:
+        // - https://identity.foundation/presentation-exchange/spec/v2.0.0/#input-evaluation
+        val inputs = getCredentials()?.filter {
+            it.status != Validator.RevocationStatus.REVOKED
+        }?.map { it.storeEntry }?.sortedBy {
+            // prefer iso credentials and sd jwt credentials over plain vc credentials
+            // -> they support selective disclosure!
+            when (it) {
+                is SubjectCredentialStore.StoreEntry.Vc -> 2
+                is SubjectCredentialStore.StoreEntry.SdJwt -> 1
+                is SubjectCredentialStore.StoreEntry.Iso -> 1
+            }
+        } ?: return KmmResult.failure(CredentialRetrievalException())
+
+        data class CandidateInputMatchContainer(
```

Comment:
I would not put a `data class` inside a method. Maybe move out some parts of this large method?

Reply:
fixed in 3e4c8cd54334e7b7ed0bb56817ae64d8ed0d8d0d